018-inertiajs.md

```diff
@@ -0,0 +1,81 @@
+
+That being said, it has had a lot of support within the Laravel community and a lot of recent commit activity. The documentation is very thorough and it even has an entire [live example demo site](https://inertiajs.com/demo-application) built using Laravel, InertiaJS and VueJS with the code available as an [open source repository](https://github.com/inertiajs/pingcrm). The community has even contributed versions of the demo site using other server-side and client-side languages/frameworks, including one using [Laravel, InertiaJS and ReactJS](https://github.com/Landish/pingcrm-react).
+
+## Alternatives
+
+There are a couple of alternatives we also considered, but with either one it would potentially mean either less flexibility, or lots of overhead and more maintenance to enable our existing backend code to communicate with the frontend.
+
+### Laravel Nova
+
+[Laravel Nova](https://nova.laravel.com/) is a one-time purchase that essentially provides a full administrative panel for a Laravel application.
+
+It has a whole slew of features that are very compelling. The main benefit Nova provides is that the team would not have to do much work building out an administrative panel.
+
+However, I am cautious that how our Eloquent models have evolved and the different relationships we have established between resources, combined with the merging of services may not completely follow the Laravel approach and we could run into numerous headaches.
+
+I also imagine we would have to wait until we complete the unified database and switch off MongoDB completely before we could move to use Nova, to avoid issues since the MongoDB Eloquent driver is a community contribution and not officially supported by Laravel.
+
+Nova does offer CLI generators to create custom tools within the framework that could support our business needs. The downside is that these tools are generated as Vue components, which would require our developers to learn [Vue JS](https://vuejs.org/) and context switch between Vue JS for administrative tools and React JS for member facing tools. While the need for custom tools may be low, it would imply additional maintenance overhead in a client-side framework the team is not familiar with.
+
+### React Admin
+
+[React Admin](https://marmelab.com/react-admin) is an open-source framework for building admin applications in React with data provided by REST/GraphQL APIs.
+
+The framework looks like is has a number of useful components to help build out an administrative interface. The main benefit React Admin provides is that the team would not have to do much work building out components for an administrative interface.
+
+However, the framework is dependent on having established API endpoints for all the resources it will render. This would require the team to invest time and effort into building out endpoints for all backend resources and maintaining them over time.
```

Comment:
Don't we already do this? I don't think there's really anything from the former Rogue at least that isn't also surfaced in the API but maybe I'm not thinking of the right example here!

Reply:
I'm into the idea of Inertia but React Admin seems cool too so I want to make sure I understand why it was ruled out, so I have a few more curiosities! 

Agreed that not all interfaces in Rogue were React and that maybe not everything was surfaced for Admins, but I think everything was surfaced via the API, whether we actually used it for the front-end or not. Chompy I am less sure about and could see being a pain point for sure. But we know everything in Aurora was available via the API since it was a separate app from Northstar. 

I see how this approach could cause extra API maintenance like if there are things we need to surface to admins that Phoenix never needs to know about - is that the sort of thing you're thinking about? Or is there a historical example where maybe using the Northstar API for both Phoenix and Aurora caused issues?

Reply:
> I see how this approach could cause extra API maintenance like if there are things we need to surface to admins that Phoenix never needs to know about - is that the sort of thing you're thinking about?

Ah yes, moreso this! Sorry if I wasn't super clear and if desired, I can amend the text to make this point clearer. 😅 

If everything is an API endpoint for both the admin frontend and the external services, there will be a number of endpoints the external services do not care about (only used for the admin frontend) and plenty where there will be overlap which both would use. This could cause problems in itself and needing to constantly feature-flag or version control endpoints if we want to change something up for one but not the other. Having an external dependency that could break when changes are made to an endpoint adds extra overhead and complexity.

Inertia essentially lets us keep all our `routes/web.php` for our admin app and `routes/api.php` for any external services, which feels like a nice separation knowing exactly where to go for what you need to do.

But, as I mention briefly in the "Combo" section, just because we use Inertia does _not_ mean we cannot use a React UI library similar to React Admin (but just UI components, without it doing its own requests for data) instead of rolling our own. I would like to explore rolling our own components initially to see how feasible it is and if we could have a collection of basic components that both Northstar Admin and Phoenix could utilize (new Forge? 🤔 ).

Reply:
Thanks Diego - I totally see where you're coming from more clearly now! Particularly liked the example where maybe we want to change something for one but not the other - having them separated definitely avoids that! I think it couldn't hurt to pop some of this reasoning into the main RFC itself but up to you!

Reply:
Sounds good! I'll amend the RFC and add this into it. Appreciate you questioning it, since I hadn't realized it wasn't super clear on that point. 🙌

Reply:
This is a great discussion, and definitely worth writing up a bit more on in the RFC! 🙌 

I think we generally created non-API admin panels (like the "create campaign" and "create action" forms) when it seemed like it'd be too much effort to wire all this up to an API endpoint & GraphQL mutation. Something I'm excited to experiment with is Inertia's support for form submission & [validation](https://inertiajs.com/validation), which would probably have made these two forms much less daunting to build in React.

Reply:
Ah those are great examples, that makes a lot of sense! I thought the form submission & validation stuff that comes with it is super cool too!